eoflow/base/base_model.py

```diff
@@ -111,3 +84,94 @@ def build_model(self, features, labels, mode):
+            in the shape required by the model.
+        :type dataset_fn: function
+        :param num_epochs: Number of epochs.
+        :type num_epochs: int
+        :param output_directory: Output directory, where the model checkpoints and summaries are saved.
+        :type output_directory: str
+        :param save_steps: Number of steps between saving model checkpoints.
+        :type save_steps: int
+        :param summary_steps: Number of steps between recodring summaries.
+        :type summary_steps: int
+        :param progress_steps: Number of steps between outputing progress to stdout.
+        :type progress_steps: int
+        """
+        
+        # Clear graph
+        tf.reset_default_graph()
+
+        with tf.Session() as sess:
+            # Build the dataset
+            dataset = dataset_fn()
+            iterator = dataset.make_initializable_iterator()
+            features, labels = iterator.get_next()
+
+            # Build model
+            self.init_global_step()
+            train_op, loss_op, summaries_op = self.build_model(features, labels, ModelMode.TRAIN)
+
+            # Create saver
+            step_tensor = self.global_step_tensor
+            checkpoint_dir = os.path.join(output_directory, 'checkpoints')
+            create_dirs([checkpoint_dir])
+            checkpoint_path = os.path.join(checkpoint_dir, 'model.ckpt')
+            saver = tf.train.Saver()
+
+            # Restore latest checkpoint if it exits
+            checkpoint_file = tf.train.latest_checkpoint(checkpoint_dir)
+            if checkpoint_file is not None:
+                print("Restoring checkpoint: %s" % checkpoint_file)
+                saver.restore(sess, checkpoint_file)
+
+            # Create summary writer
+            create_dirs([checkpoint_dir])
+            summary_writer = tf.summary.FileWriter(output_directory, sess.graph)
+
+            # Initialize variables
+            initializer = tf.global_variables_initializer()
+            sess.run(initializer)
+
+            # Train
+            try:
+                training_step = 1
+                for e in range(num_epochs):
+                    sess.run(iterator.initializer)
+                    print("Epoch %d/%d" % (e+1, num_epochs))
+
+                    while True:
+                        try:
+                            # Compute and record summaries every summary_steps
+                            if training_step % summary_steps == 0:
+                                _, loss, step, summaries = sess.run([train_op, loss_op, step_tensor, summaries_op])
+
+                                summary_writer.add_summary(summaries, global_step=step)
+                            else:
+                                _, loss, step = sess.run([train_op, loss_op, step_tensor])
+
+                            # Show progress
+                            if training_step % progress_steps == 0:
```

Comment:
it would be nice to use a progress bar such as the one provided by `tqdm`, and adding the loss value to the progress bar using `pbar.set_description`